tools/GSDumpGUI/Library/GSDXWrapper.cs

```diff
@@ -205,115 +205,135 @@ public unsafe void Run(GSDump dump, int rendererOverride)
+                                {
+                                    switch(msg.wParam.ToInt32() & 0xFF)
+                                    {
+                                        case 0x1B: Running = false; break; // VK_ESCAPE;
+                                        case 0x77: GSmakeSnapshot(""); break; // VK_F8;
+                                    }
+                                }
                             }
 
-                            GSreset();
-                            Marshal.Copy(dump.Registers, 0, new IntPtr(pointer), 8192);
-                            GSsetBaseMem(new IntPtr(pointer));
-                            GSfreeze(0, new IntPtr(fr));
+                            if (!Running || !NativeMethods.IsWindowVisible(hWnd))
+                                break;
 
-                            for (int i = 0; i < dump.Data.Count; i++)
+                            if (DebugMode)
                             {
-                                GSData itm = dump.Data[i];
-                                CurrentGIFPacket = itm;
-
-                                if (DebugMode)
+                                if (QueueMessage.Count > 0)
                                 {
-                                    if (RunTo != -1)
+                                    TCPMessage Mess = QueueMessage.Dequeue();
+                                    switch (Mess.MessageType)
                                     {
-                                        if (i == RunTo)
-                                        {
-                                            RunTo = -1;
-                                            int idxnextReg = dump.Data.FindIndex(i, a => a.id == GSType.Registers);
-                                            if (idxnextReg != -1)
-                                            {
-                                                Step(dump.Data[idxnextReg], pointer);
-                                            }
-
-                                            GSData g = new GSData();
-                                            g.id = GSType.VSync;
-                                            Step(g, pointer);
-
-                                            TCPMessage Msg = new TCPMessage();
-                                            Msg.MessageType = MessageType.RunToCursor;
-                                            Msg.Parameters.Add(i);
-                                            Program.Client.Send(Msg);
-
-                                            ExternalEvent.Set();
-                                        }
-                                        else
-                                        {
-                                            Step(itm, pointer);
-                                        }
+                                        case MessageType.Step:
+                                            if (debug_idx >= dump.Data.Count) debug_idx = 0;
```

Comment:
Also in all of these I’d say newlines improve readability

Reply:
Done! I think.